source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,149 @@
+#include "opt/pass.h"
+
+#include <queue>
+#include <utility>
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status LICMPass::Process(ir::IRContext* context) {
+  InitializeProcessing(context);
+  bool modified = false;
+
+  if (context != nullptr) {
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = get_module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor* loop_descriptor = context()->GetLoopDescriptor(f);
+
+  // Process each loop in the function
+  for (ir::Loop& loop : *loop_descriptor) {
+    // Ignore nested loops, as we will process them in order in ProcessLoop
+    if (loop.IsNested()) {
+      continue;
+    }
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  bool modified = false;
+
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    modified |= ProcessLoop(nested_loop, f);
+  }
+
+  // Get the pre_header. If it doesn't exist, force it's creation
+  ir::BasicBlock* pre_header_bb = loop->GetPreHeaderBlock();
+  if (!pre_header_bb) {
+    modified = true;
+    pre_header_bb = loop->GetOrCreatePreHeaderBlock();
+  }
+  ir::BasicBlock* header_bb = loop->GetHeaderBlock();
+  ir::BasicBlock* merge_bb = loop->GetMergeBlock();
+  std::queue<ir::BasicBlock*> loop_bbs{};
+  modified |=
+      AnalyseAndHoistFromBB(loop, f, header_bb, merge_bb,
+                            pre_header_bb, loop->GetHeaderBlock(), &loop_bbs);
+
+  while (!loop_bbs.empty()) {
+    ir::BasicBlock* bb = loop_bbs.front();
+    loop_bbs.pop();
+    modified |= AnalyseAndHoistFromBB(loop, f, header_bb,
+                                      merge_bb, pre_header_bb, bb, &loop_bbs);
+  }
+
+  return modified;
+}
+
+bool LICMPass::AnalyseAndHoistFromBB(ir::Loop* loop, ir::Function* f,
+                                     ir::BasicBlock* header_bb,
+                                     ir::BasicBlock* merge_bb,
+                                     ir::BasicBlock* pre_header_bb,
+                                     ir::BasicBlock* bb,
+                                     std::queue<ir::BasicBlock*>* loop_bbs) {
+  bool modified = false;
+  if (!IsInsideNestedLoop(loop, f, bb)) {
+    bool hoisted = true;
+    while (hoisted) {
+      hoisted = false;
+      // When we perform a hoist it will invalidate this iterator, but we
+      // require the hoist so instructions later in the bb which depended on the
+      // hoisted instruction can now be hoisted. As such, we break iteration
+      // over the bb and start again with fresh iterators until no instructions
+      // are hoisted
```

Comment:
You can get around this by not using iterators.  The `ForEachInst` and `WhileEachInst` are written to handle deleting or moving the current instruction.  We should add that to the documentation for them because we rely on it in a few places.

Can be done in a follow up PR.